packages/autocomplete-plugin-tags/src/createTags.ts

```diff
@@ -0,0 +1,72 @@
+import { BaseItem } from '@algolia/autocomplete-core';
+import { createRef } from '@algolia/autocomplete-shared';
+
+import { CreateTagsPluginParams } from './createTagsPlugin';
+import { BaseTag, Tag } from './types';
+
+export type OnTagsChangeParams<TTag> = {
+  prevTags: Array<Tag<TTag>>;
+  tags: Array<Tag<TTag>>;
+};
+
+type OnTagsChangeListener<TTag> = (params: OnTagsChangeParams<TTag>) => void;
+
+type CreateTagsParams<TTag, TItem extends BaseItem> = Pick<
+  CreateTagsPluginParams<TTag, TItem>,
+  'initialTags'
+>;
+
+export function createTags<TTag, TItem extends BaseItem>({
+  initialTags = [],
+}: CreateTagsParams<TTag, TItem>) {
+  const tagsRef = createRef(toTags(initialTags));
+  const onChangeListeners: Array<OnTagsChangeListener<TTag>> = [];
+
+  function toTags<TTag>(baseTags: Array<BaseTag<TTag>>): Array<Tag<TTag>> {
+    return baseTags.map((baseTag) => {
+      const tag = {
+        ...baseTag,
+        remove() {
+          const prevTags = tagsRef.current.slice();
+          const newTags = tagsRef.current.filter(
+            (tagRef) => tag !== ((tagRef as unknown) as Tag<TTag>)
```

Comment:
why does this need to be casted? is initialTags not already Array<Tag<TTag>>?

Reply:
It should, but TypeScript doesn't understand it somehow.

<img width="511" alt="Capture d’écran 2021-08-24 à 15 53 26" src="https://user-images.githubusercontent.com/5370675/130629258-5c8561cf-1f06-4fc8-acc6-917915ae81ec.png">

It seems to be a [known issue](https://github.com/microsoft/TypeScript/issues/25642#issuecomment-404929935) that this rule has false positives.

Reply:
Isn't it because we compare it to `tags` which is being initialized?

Reply:
No, this does the same if you create a dummy tag before.

<img width="511" alt="Capture d’écran 2021-08-24 à 15 53 26" src="https://user-images.githubusercontent.com/5370675/130654043-0127ac81-d73f-42c2-baf6-c8020744fedc.png">

The type of `tag` and `tagRef` are properly inferred, even without type casting:

<img width="240" alt="Capture d’écran 2021-08-24 à 18 27 44" src="https://user-images.githubusercontent.com/5370675/130654353-4f6f568d-f7b1-4872-b3d0-be93f8cd37f6.png">
<img width="268" alt="Capture d’écran 2021-08-24 à 18 28 03" src="https://user-images.githubusercontent.com/5370675/130654366-47605f77-e37a-4bb0-9312-28a2a5b0d94d.png">

I don't think it's a type issue per se, because we're type casting `tagRef` to what it's already inferred as. My guess is that it's a bug in TypeScript which is inconsistent between what it says it infers as and what errors it raises. I can open an issue in the TS repo but with 4.9K current issues, I don't think this is worth it.